addons/udes_stock_cron/models/stock_picking.py

```diff
@@ -0,0 +1,387 @@
+# -*- coding: utf-8 -*-
+from odoo import fields, models, _, api
+from collections import defaultdict
+from odoo.exceptions import UserError, ValidationError
+from psycopg2 import OperationalError, errorcodes
+
+import logging
+_logger = logging.getLogger(__name__)
+
+
+PG_CONCURRENCY_ERRORS_TO_RETRY = (
+    errorcodes.LOCK_NOT_AVAILABLE,
+    errorcodes.SERIALIZATION_FAILURE,
+    errorcodes.DEADLOCK_DETECTED,
+)
+MAX_TRIES_ON_CONCURRENCY_FAILURE = 5
+
+class StockPicking(models.Model):
+    _inherit = 'stock.picking'
+
+    batch_id = fields.Many2one("stock.picking.batch", copy=True)
+
+    def _reserve_stock_assign(self):
+        """Perform assign of the pickings at the move level because refactoring
+        may change the pickings.
+        This is a function so it can be extended and/or overriden.
+        """
+        # Assign at the move level disabling refactoring
+        moves = self.mapped("move_lines")
+        moves.with_context(lock_batch_state=True, disable_move_refactor=True)._action_assign()
+
+        # Unreserve any partially reserved lines if not allowed by the picking type
+        moves._unreserve_partial_lines()
+
+        # Refactor after unreserving
+        refactored_moves = moves._action_refactor(stage="assign")
+
+        return refactored_moves.mapped("picking_id")
+
+    def reserve_stock(self):
+        """
+        Reserve stock according to the number of reservable pickings.
+        If this method is called on an empty recordset it will attempt to
+        reserve stock for all eligible picking types.  If the recordset is not
+        empty, it will reserve stock for the picks in the recordset. If the
+        recordset is not empty, it is the callers responsibility to make sure
+        that the pickings belong to at most one batch, otherwise this method
```

Comment:
What is the issue with if self is populated, raising an error when the pickings in self are in more than one batch?